### PR TITLE
[2/X] [Ready for review] Add servlet for signup form.

### PR DESCRIPTION
**Pull request by nikhitashanker**
- Add servlet for volunteering opportunity signup form
- Add tests

### CODE REVIEW

**Comment by nikhitashanker** [comment]
[Presubmit errors to be fixed by events PRs]

File: project/src/main/java/com/google/sps/servlets/OpportunitySignupFormHandlerServlet.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+   */
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String opportunityId = request.getParameter(OPPORTUNITY_ID);
+    User user = UserServiceFactory.getUserService().getCurrentUser();
+
+    if (opportunityId == null) {
+      response.sendError(HttpServletResponse.SC_UNAUTHORIZED, "Opportunity ID not specified.");
+    } else if (user == null) {
+      response.sendError(HttpServletResponse.SC_UNAUTHORIZED, "You have not logged in.");
```
Comment (yicyao) [inline project/src/main/java/com/google/sps/servlets/OpportunitySignupFormHandlerServlet.java:34]: can we redirect them to login?

File: project/src/test/java/com/google/sps/OpportunitySignupFormHandlerTest.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+  @After
+  public void tearDown() {
+    SpannerTestTasks.cleanup();
+    helper.tearDown();
+  }
+
+  @Test
+  public void testAddOpportunitySignup_LoggedIn() throws IOException {
+    VolunteeringOpportunity opportunity =
```
Comment (liangqigong) [inline project/src/test/java/com/google/sps/OpportunitySignupFormHandlerTest.java:51]: For each line of this part, please make such a decision: If the reader doesn't see this line, can the reader still understand the test.

If the answer is yes, move the line to setUp.

File: project/src/test/java/com/google/sps/OpportunitySignupFormHandlerTest.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        new VolunteeringOpportunity.Builder(EVENT_ID, NAME, NUMBER_OF_SPOTS).build();
```
Comment (liangqigong) [inline project/src/test/java/com/google/sps/OpportunitySignupFormHandlerTest.java:52]: Event_id, name, number_spots don't affect the test. So move 

new VolunteeringOpportunity.Builder(EVENT_ID, NAME, NUMBER_OF_SPOTS).build()

to function newVolunteeringOpportunity().

File: project/src/test/java/com/google/sps/OpportunitySignupFormHandlerTest.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+    new OpportunitySignupFormHandlerServlet().doPost(request, response);
+
+    Mockito.verify(response).sendRedirect("/event-details.html");
+  }
+
+  @Test
+  public void testAddOpportunitySignup_OpportunityIdNotSpecified() throws IOException {
+    HttpServletRequest request = Mockito.mock(HttpServletRequest.class);
+    HttpServletResponse response = Mockito.mock(HttpServletResponse.class);
```
Comment (liangqigong) [inline project/src/test/java/com/google/sps/OpportunitySignupFormHandlerTest.java:70]: For this test, the reader have no idea why it throws "Opportunity ID not specified."

So add one line like this:
    Mockito.when(request.getParameter(OPPORTUNITY_ID)).thenReturn(null);

File: project/src/main/java/com/google/sps/servlets/OpportunitySignupFormHandlerServlet.java (hunk 1 above)
Comment (nikhitashanker) [reply project/src/main/java/com/google/sps/servlets/OpportunitySignupFormHandlerServlet.java:34]: Done - good idea.

File: project/src/test/java/com/google/sps/OpportunitySignupFormHandlerTest.java (hunk 4 above)
Comment (nikhitashanker) [reply project/src/test/java/com/google/sps/OpportunitySignupFormHandlerTest.java]: Done - I see, I will keep this in mind.

File: project/src/main/java/com/google/sps/servlets/OpportunitySignupFormHandlerServlet.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+   * @param response servlet response
+   * @throws IOException if Input/Output error occurs
+   */
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String opportunityId = request.getParameter(OPPORTUNITY_ID);
+    User user = UserServiceFactory.getUserService().getCurrentUser();
+
+    if (opportunityId == null) {
+      response.sendError(HttpServletResponse.SC_UNAUTHORIZED, "Opportunity ID not specified.");
```
Comment (dash102) [inline project/src/main/java/com/google/sps/servlets/OpportunitySignupFormHandlerServlet.java:32]: SC_UNAUTHORIZED might not be the most appropriate for this case, maybe a 400?

File: project/src/test/java/com/google/sps/OpportunitySignupFormHandlerTest.java (hunk 3 above)
Comment (nikhitashanker) [reply project/src/test/java/com/google/sps/OpportunitySignupFormHandlerTest.java]: Done - moved to a helper method newVolunteeringOpportunity in TestUtils.

File: project/src/test/java/com/google/sps/OpportunitySignupFormHandlerTest.java (hunk 2 above)
Comment (nikhitashanker) [reply project/src/test/java/com/google/sps/OpportunitySignupFormHandlerTest.java]: Done - thanks.

File: project/src/main/java/com/google/sps/servlets/OpportunitySignupFormHandlerServlet.java (hunk 5 above)
Comment (nikhitashanker) [reply project/src/main/java/com/google/sps/servlets/OpportunitySignupFormHandlerServlet.java:32]: Done - yes I think SC_BAD_REQUEST is more appropriate as well.

File: project/src/main/java/com/google/sps/servlets/OpportunitySignupFormHandlerServlet.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+import java.io.IOException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+ 
+/** Servlet that handles submission of the form for signing up for opportunities. */
+@WebServlet("/opportunity-signup-form-handler")
+public class OpportunitySignupFormHandlerServlet extends HttpServlet {
+  private static final String OPPORTUNITY_ID = "opportunity-id";
```
[the remark this answers is not in the document]
Comment (nikhitashanker) [reply project/src/main/java/com/google/sps/servlets/OpportunitySignupFormHandlerServlet.java:17]: Done.

**Pull request merged**